### PR TITLE
Cache cargo binaries

### PR DESCRIPTION
**Pull request by lzutao**
changelog: none

### CODE REVIEW

**Comment by lzutao** [comment]
This should cut build time in half.
r? @flip1995

File: .travis.yml [hunk 1]
```diff
@@ -122,13 +134,10 @@ script:
     fi
 
 after_success: |
-  #!/bin/bash
   if [ "$TRAVIS_OS_NAME" == "linux" ]; then
-    set -ex
```
Comment (lzutao) [inline .travis.yml]: This should have -e. Will fix.

**Comment by lzutao** [comment]
Hm, cannot install master toolchain on osx. Investigating!

**Comment by lzutao** [comment]
Reported upstream: https://github.com/kennytm/rustup-toolchain-install-master/issues/28

**Comment by flip1995** [comment]
I also tried this once. But the caches this builds are >500MB big. I'm not sure if this is a trade-off we want to / can take.

**Comment by phansch** [comment]
So, we should be able to easily test this: The first try run is using an empty cache, because there's no cache for the `try` branch yet:

```
attempting to download cache archive
fetching try/cache-osx-7170efeb401a81a9aa99b33477d640a2606564208e3f17dea84058061c3ec1a5.tgz
fetching master/cache-osx-7170efeb401a81a9aa99b33477d640a2606564208e3f17dea84058061c3ec1a5.tgz
could not download cache
```
(https://travis-ci.com/rust-lang/rust-clippy/jobs/246547175#L71)


The second try run should be faster then

**Comment by phansch** [comment]
Full cache _upload_ took 15-20 seconds in the various jobs.

@bors try

**Comment by phansch** [comment]
The cache download seems to take less than 10 seconds in every case. Less than 5 in most cases. Let's see the total time savings :hourglass_flowing_sand:

**Comment by phansch** [comment]
The second try run was ~5 minutes faster. Not sure if that's within the normal variance. Let's `try` again

@bors try

**Comment by flip1995** [comment]
So the build time for the base tests is about 5\~7 minutes faster, which is about 18\~25%. I don't think that this will make much of a difference on `try` and `r+` runs, since they are mostly limited by the limited amount of concurrent jobs. What I like most about this is that almost all integration tests are under 10 minutes and around 7~8 minutes. I think that this will improve the running time for `try` and `r+` builds the most.

Sadly it doesn't display the `Ran for` time, because the windows build got interrupted/never started.

The total time is about 35\~37% faster though.

**Comment by phansch** [comment]
Let's get this merged, it's a good improvement in any case!

@bors r+

**Pull request merged**